File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.py
def solution(today, terms, privacies):
    answer = []
    
    today = list(map(int, today.split(".")))
    
    # terms -> term_dict
    term_dict = {}
    for t in terms:
        tmp_arr = t.split(" ")
        term_dict[tmp_arr[0]] = int(tmp_arr[1])
    
   # privacies split
    for (index, privacy) in enumerate(privacies):
        tmp_arr = privacy.split(" ")
        contract_date = list(map(int, tmp_arr[0].split(".")))
        terms_type = tmp_arr[1]
        
        print("b:   ", today, contract_date, terms_type)
        
        # check possible date
        contract_date[1] += term_dict[terms_type]
        if contract_date[1] > 12:
            q, r = divmod(contract_date[1], 12)
            if r == 0:
                contract_date[1] = 12
                contract_date[0] += (q - 1)
            else:
                contract_date[1] = r
                contract_date[0] += q
        
        contract_date[2] -= 1
        if contract_date[2] == 0:
            contract_date[2] = 28
            contract_date[1] -= 1
            if contract_date[1] == 0:
                contract_date[1] = 12
                contract_date[0] -= 1
        
        # check answer
        if today[0] > contract_date[0]:
            answer.append(index + 1)
        elif today[0] == contract_date[0] and today[1] > contract_date[1]:
            answer.append(index + 1)
        elif today[0] == contract_date[0] and today[1] == contract_date[1] and today[2] > contract_date[2]:
            answer.append(index + 1)
    
        print("a:   ", today, contract_date, terms_type)
    return answer

File: 프로그래머스/unrated/150370. 개인정보 수집 유효기간/개인정보 수집 유효기간.swift
import Foundation

func solution(_ today: String, _ terms: [String], _ privacies: [String]) -> [Int] {
    var answer = [Int]()
    
    let now = today.components(separatedBy: ".").map{ Int($0)! }
    
    let termsComponents = terms.map{ $0.components(separatedBy: " ") }
    var termsDictionary = [String:Int]()
    termsComponents.map{ termsDictionary[$0[0]] = Int($0[1]) }
    
    let privaciesComponents = privacies.map{ $0.components(separatedBy: " ") }
    
    for (idx, p) in privaciesComponents.enumerated() {
        let end = p[0].components(separatedBy: ".").map{ Int($0)! }
        let type = p[1]
        
        var (y, m, d) = (end[0], end[1], end[2])
                
        m += termsDictionary[type]!
        if m > 12 {
            let (q, r) = m.quotientAndRemainder(dividingBy: 12)
            if r == 0 {
                m = 12
                y += (q - 1)
            } else {
                m = r
                y += q
            }
        }
        
        d -= 1
        if d == 0 {
            d = 28
            m -= 1
            if m == 0 {
                m = 12
                y -= 1
            }
        }
        
        if now[0] > y {
            answer.append(idx + 1)
        } else if now[0] == y, now[1] > m {
            answer.append(idx + 1)
        } else if now[0] == y, now[1] == m, now[2] > d {
            answer.append(idx + 1)
        }

    }
        
    return answer
}